xml/System.Security.Cryptography.X509Certificates/X509VerificationFlags.xml

```diff
@@ -131,7 +131,7 @@
       </ReturnValue>
       <MemberValue>16</MemberValue>
       <Docs>
-        <summary>Ignore that the chain cannot be verified due to an unknown certificate authority (CA).</summary>
+        <summary>Ignore that the chain cannot be verified due to an unknown certificate authority (CA), or a partial chain. Note that ignoring partial chain skips verification that the signing authority of the certificate under validation was actually signed by one of the known CAs, so this must be performed separately if desired.</summary>
```

Comment:
I'm not sure what the second sentence is trying to say.  If a candidate is found but the signature doesn't work out, that produces a different error.

```suggestion
        <summary>Ignore that the chain cannot be verified due to an unknown certificate authority (CA), or a partial chain.</summary>
```

Reply:
I'm trying to elaborate on the consequences of not validating partial chain - in particular, that because partial chain status gets suppressed, the certificate under validation is essentially always validated:

If unknown CAs are permitted, the certificate under validation doesn't actually have to be signed by any of the certificates added to the trust store for validation to still succeed.

IMO this is a key piece of information the docs need to disclose; that users should expect to have to do their own validation of the fingerprint on the end-certificate's issuer and signing root to ensure those values are one and the same.

Reply:
> If unknown CAs are allowed, the certificate under validation doesn't actually have to be signed by any of the certificates added to the trust store for validation to still succeed. 

Yes, that's true by definition: By specifying this flag you've said it's OK for the chain to end in a CA you don't know about.  Whether last thing in the chain is a root authority or an intermediate (or the end-entity itself) you've said "trust doesn't matter" and you need to do something about it.

Reply:
> true by definition: By specifying this flag you've said it's OK for the chain to end in a CA you don't know about.

The docs don't describe this definition. They say it only ignores `UntrustedRoot` flag, which is the behavior most users are actually after; verifying a complete chain but allowing for the `UntrustedRoot` chain status flag is very common for validating certificates against a self-signed root CA. `Build()` returning verification success for any chain at all in this scenario has not been documented nor is expected.

Most other software that interacts with X.509 has similar flags or options for validating a certificate with an unknown/self-signed certificate authority, but they otherwise keeps all rules the same - i.e. requiring completed chain, and hence the user confusion around this.

The .NET implementation's unexpected behavior regarding partial chain and verification success has come up in the past (dotnet/runtime/issues/26449) and [has turned users away](https://stackoverflow.com/a/32872136).

What do you suggest?

Reply:
> verifying a complete chain but allowing for the UntrustedRoot chain status flag is very common for validating certificates against a self-signed root CA

The correct piece of action for that scenario is that after calling Build you check that `chain.ChainElements[^1].Certificate.RawData.SequenceEquals(s_pinnedCertBytes)`.  If you're doing that, it doesn't matter if it ended in UnknownRoot or PartialChain (or a known/trusted root).  If you're not doing that, you're not actually testing it came from your own root, and is a security hole.

The only way that the two interact is if you did a check that if the last one ended in UntrustedRoot then you check pinning (which probably isn't what's desired, but maybe it is during "I plan to use a trusted CA cert soon" rollout phases), in which case merely the "or a partial chain" addition in docs covers the scenario.

Reply:
I don't disagree with that, and it's been described well in the above linked GitHub issue discussions which is why I'm trying to move it into the docs. If you feel the examples section is a better place for it, I'm happy to add one there.

But let's put specific root CA pinning aside, and consider a more general case of a user wanting to use `chain.Build()` to verify a given certificate against the OS trust store --

> in which case merely the "or a partial chain" addition in docs covers the scenario

IMO "or a partial chain" is a far cry from its actual implications, namely that the user would **always** need to perform manual issuer verification, because ignoring partial chain implies `X509Chain.Build()` now returns `true` no matter what the contents of the trusted certificate store is. That behavior is unintuitive and these classes are in a security sensitive area, so I think it warrants being explicit for the sake of our users.

If you feel that the docs only warrant a shorter "or a partial chain" then let me know and I'll adjust my PR.

Reply:
> But let's put specific root CA pinning aside, and consider a more general case of a user wanting to use chain.Build() to verify a given certificate against the OS trust store

Then they don't set the `X509VerificationFlags.AllowUnknownCertificateAuthority` flag, and now `PartialChain` (or an unknown root) make `Build()` return `false`, and everything's happy.

>  If you feel that the docs only warrant a shorter "or a partial chain" then let me know

I do feel that way.  Especially because of the technical errors in your statement (e.g. it'll return false if it's expired (unless the "ignore validity times" flag is also set)).

Reply:
> Especially because of the technical errors in your statement (e.g. it'll return false if it's expired (unless the "ignore validity times" flag is also set)).

Fair enough, the first sentence of the elaboration was ambiguous. Either way, I've removed all of it at your request.